File: Python/0300-m-longest-increasing-subsequence.py
"""
DP
F[1] = 1
F[i] = max{1, F[j]+1 | aj<ai and j < i}
Time: O(n^2)
Space: O(n)
"""
class Solution_1:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if not nums:
            return 0
        n = len(nums)
        dp = [1] * n
        for curt in range(n):
            for prev in range(curt):
                if nums[curt] > nums[prev]:
                    dp[curt] = max(dp[curt], dp[prev] + 1)
        return max(dp)

"""
Binary Search
Time: O(nlogn)
Space: O(n)
"""
class Solution_2:
    def lengthOfLIS(self, nums: List[int]) -> int:
        res = []
        for x in range(len(nums)):
            start = 0
            end = len(res) - 1
            while start <= end:
                mid = start + (end - start) // 2
                if res[mid] < nums[x]:
                    start = mid + 1
                else:
                    end = mid - 1
            if start >= len(res):
                res.append(nums[x])
            else:                               
                res[start] = nums[x]
        return len(res)

File: leetcode/C++/0300-m-longest-increasing-subsequence.cpp
// DP
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0;
        vector<int> f(n, 1);
        int res = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; ++j) {
                if (nums[j] < nums[i])
                    f[i] = max(f[i], f[j] + 1);
            }
            res = max(res, f[i]);
        }
        return res;
    }
};